### ISSUE TITLE
ompi_datatype_is_contiguous_memory_layout

### ISSUE THREAD

**Issue by wkliao** [open-mpi/ompi#6919]
A question about the behavior of function `ompi_datatype_is_contiguous_memory_layout()`
called in `ADIOI_Datatype_iscontig()` in file `ompi/mca/io/romio321/src/io_romio321_module.c`

I tested this function with a zero-length datatype, for example,
```
MPI_Type_create_hvector(0, 10, 2, MPI_INT, &dtype);
```
and it returns 0. meaning dtype is not contiguous. Is this expected?

On the other hand, I also check MPICH implementation of `ADIOI_Datatype_iscontig()`
which  calls `MPIR_Datatype_iscontig()`. It returns 1 for the same 0-length hvector datatype.

Because of this different behavior, the flattened file views, in term of offset-length lists, are different between OpenMPI and MPICH.

CC @roblatham00

**Comment by bosilca** [comment]
Indeed, the OPAL iscontig function treats empty datatypes as non-contiguous, assuming you should check upstream if you are supposed to do something with such a datatype before asking if it is contiguous. I would add a test to check if the data has some content before returning.

```diff
diff --git a/ompi/mca/io/romio321/src/io_romio321_module.c b/ompi/mca/io/romio321/src/io_romio321_module.c
index 0f7016e..f580a8d 100644
--- a/ompi/mca/io/romio321/src/io_romio321_module.c
+++ b/ompi/mca/io/romio321/src/io_romio321_module.c
@@ -147,5 +147,9 @@ void ADIOI_Datatype_iscontig(MPI_Datatype datatype, int *flag)

         if (true_lb > 0)
             *flag = 0;
+    } else {
+        size_t size;
+        ompi_datatype_size(datatype, &size);
+        *flag = (size ? 1 : 0);
     }
 }
```

**Comment by wkliao** [comment]
```
+        *flag = (size ? 1 : 0);
```
or 
```
+        *flag = (size ? 0 : 1);
```

**Comment by jsquyres** [comment]
Being explicit is always a Good Thing.  :-)

```c
*flag = (size == 0 ? ....);
```

That way there's no confusion.  😉

**Comment by bosilca** [comment]
@wkliao 👍

**Comment by wkliao** [comment]
Thanks. I assume a patch will be created and tested?

Also, I guess the function to get the size of a  datatype is `ompi_datatype_type_size`.
I got a link error for `ompi_datatype_size ` not found.

**Comment by bosilca** [comment]
ok, let me try again ;)

```diff
diff --git a/ompi/mca/io/romio321/src/io_romio321_module.c b/ompi/mca/io/romio321/src/io_romio321_module.c
index 0f7016e..f580a8d 100644
--- a/ompi/mca/io/romio321/src/io_romio321_module.c
+++ b/ompi/mca/io/romio321/src/io_romio321_module.c
@@ -147,5 +147,9 @@ void ADIOI_Datatype_iscontig(MPI_Datatype datatype, int *flag)

         if (true_lb > 0)
             *flag = 0;
+    } else {
+        size_t size;
+        ompi_datatype_type_size(datatype, &size);
+        *flag = (0 == size ? 1 : 0);
     }
 }
```
